Clean up Sphinx warnings related to models

jessica-mitchell opened the pull request:
This PR resolves some of the warnings that Sphinx outputs related to models.

Many of the changes are related to items in the bibliography that were not cited properly.
There are also syntax fixes.

Because the restructured text files for models are autogenerated from a script, Sphinx warns that it doesnt' have them indexed.
So this PR also indexes the files by creating a JSON file from the extractor_userdocs.py that Sphinx reads and outputs in Jinja templated table of contents.

Fixes #1907

terhorstd commented on models/cont_delay_synapse.h:
```diff
@@ -45,8 +45,8 @@ Description
 which are an integer multiple of the time step h. A continuous delay is
-decomposed into an integer part (delay_) and a double (delay_offset_) so
-that the actual delay is given by  delay_*h - delay_offset_. This can be
+decomposed into an integer part (delay\_) and a double (delay_offset\_) so
+that the actual delay is given by  delay_*h - delay_offset\_. This can be
```
Not exactly sure what is meant here, but if the underscore is code maybe like this (?)
```suggestion
decomposed into an integer part (``delay_``) and a double (``delay_offset_``) so
that the actual delay is given by  ``delay_*h - delay_offset_``. This can be
```

terhorstd commented on models/stdp_synapse_facetshw_hom.h:
```diff
@@ -47,11 +47,11 @@ stdp_facetshw_synapse is a connector to create synapses with spike-timing
 details see [3]_.
-The modified spike pairing scheme requires the calculation of tau_minus_
-within this synapse and not at the neuron site via Kplus_ like in
+The modified spike pairing scheme requires the calculation of tau_minus\_
+within this synapse and not at the neuron site via Kplus\_ like in
```
```suggestion
The modified spike pairing scheme requires the calculation of ``tau_minus_``
within this synapse and not at the neuron site via ``Kplus_`` like in
```

terhorstd commented on models/stdp_synapse_facetshw_hom.h:
```diff
@@ -86,7 +86,7 @@ Parameters
  configbit_0            list of     Configuration bits for evaluation
                         integers    function. For details see code in
-                                    function eval_function_ and [4]_
+                                    function eval_function\_ and [4]_
```
```suggestion
                                    function ``eval_function_`` and [4]_
```

terhorstd commented on models/stdp_triplet_synapse.h:
```diff
@@ -44,11 +44,12 @@ stdp_triplet_synapse is a connection with spike time dependent
   tau_plus_triplet, respectively.
 - Postsynaptic traces o_1 and o_2 of [1]_ are acquired from the postsynaptic
-  neuron states Kminus_ and triplet_Kminus_ which decay on time-constants
+  neuron states Kminus\_ and triplet_Kminusi_ which decay on time-constants
```
Also in the lines around this one
```suggestion
  neuron states ``Kminus_`` and ``triplet_Kminusi_`` which decay on time-constants
```

terhorstd commented on models/correlation_detector.h:
```diff
@@ -129,22 +129,8 @@ Correlation detectors IGNORE any connection delays.
-See Auto- and crosscorrelation functions for spike
-trains[cross_check_mip_corrdet.py]
-in pynest/examples.
+of State\_, but are initialized by init\_buffers\_().
```
```suggestion
of ``State_``, but are initialized by ``init_buffers_()``.
```

terhorstd commented on models/correlomatrix_detector.h:
```diff
@@ -88,16 +88,8 @@ sense.
-   be iterator since that may change
- - update() deletes all entries before now-tau_max, sorts the new
-   entries, then registers new entries in histogram
+ of State\_, but are initialized by init\_buffers\_().
```
```suggestion
 of ``State_``, but are initialized by ``init_buffers_()``.
```

terhorstd commented on models/correlospinmatrix_detector.h:
```diff
@@ -78,7 +77,7 @@ Correlospinmatrix detectors IGNORE any connection delays.
 Correlospinmatrix detector breaks with the persistence scheme as
 follows: the internal buffers for storing spikes are part
-of State_, but are initialized by init_buffers_().
+of State\_, but are initialized by init_buffers_().
```
```suggestion
of ``State_``, but are initialized by ``init_buffers_()``.
```

terhorstd reviewed (changes requested):
Amazingly many details fixed! Thanks a lot!
Some double-backtics need a look, as discussed.

jessica-mitchell commented:
@terhorstd I addressed your comments, but since this was updating the double back ticks for some of the terms, I decided to go through and try find all terms that needed double back ticks and add them.

terhorstd commented on models/correlomatrix_detector.h:
```diff
@@ -49,18 +49,18 @@ Device for measuring the covariance matrix from several inputs
 used to weight every count, which is available under the key /covariance.
-Both are matrices of size N_channels x N_channels, with each entry C_ij being
-a vector of size tau_max/delta_tau + 1 containing the (weighted) histogram
+Both are matrices of size :math:`N_channels x N_channels`, with each entry :math:`C_ij` being
```
How does `:math:` handle the subscript? Could you check that `j` is also part of it? In LaTeX notation this would require curly braces `C_{ij}` to not end up as `C_{i} j`…

jessica-mitchell replied:
Fixed these

terhorstd commented on models/gamma_sup_generator.h:
```diff
@@ -61,6 +61,9 @@ gamma_shape
 n_proc
```
Should these also be double backticked? Depends also on the style in the `include`d description.

terhorstd replied:
I saw this also in many later files. Do not fix this in this here though, we need a general approach to this. Just fixing this in all files could easily be its own PR.

terhorstd commented on models/gauss_rate.h:
```diff
@@ -49,24 +49,26 @@ Rate neuron model with Gaussian gain function
 Nonlinear rate neurons can be created by typing
-nest.Create('gauss_rate_ipn'). Nonlinear rate transformers can be
-created by typing nest.Create('rate_transformer_gauss').
+nest.Create(``gauss_rate_ipn``). Nonlinear rate transformers can be
+created by typing ``nest.Create('rate_transformer_gauss``).
```
```suggestion
``nest.Create("gauss_rate_ipn")``. Nonlinear rate transformers can be
created by typing ``nest.Create("rate_transformer_gauss")``.
```

terhorstd commented on models/glif_cond.h:
```diff
@@ -52,14 +52,14 @@ Conductance-based generalized leaky integrate and fire (GLIF) model
 Incoming spike events induce a postsynaptic change of conductance modeled
 by an alpha function [2]_. The alpha function is normalized such that an event
-of weight 1.0 results in a peak conductance change of 1 nS at t = tau_syn. On
+of weight 1.0 results in a peak conductance change of 1 nS at `t = tau_syn`. On
```
```suggestion
of weight 1.0 results in a peak conductance change of 1 nS at ``t = tau_syn``. On
```

terhorstd replied:
(see comments below first!)

terhorstd commented on models/glif_cond.h:
```diff
@@ -130,36 +130,36 @@ V_reset    double   Reset potential of the membrane in mV (GLIF 1 or GLIF 3)
 -------------------------------------------------------------------------------
 th_spike_add               double         Threshold addition following spike
```
Same question about parameter names here. Do they require backticks?

jessica-mitchell replied:
I think parameters in tables don't necessarily need them. If we want to add them, it would take a fair bit of effort because the tables need to be realigned.

terhorstd replied:
fine with me.

terhorstd commented on models/glif_psc.h:
```diff
@@ -42,11 +42,11 @@ Current-based generalized leaky integrate-and-fire models
 Incoming spike events induce a postsynaptic change of current modeled
 by an alpha function [2]_. The alpha function is normalized such that an event
-of weight 1.0 results in a peak current of 1 pA at t = tau_syn. By default,
+of weight 1.0 results in a peak current of 1 pA at `t = tau_syn`. By default,
```
```suggestion
of weight 1.0 results in a peak current of 1 pA at ``t = tau_syn``. By default,
```

terhorstd commented on models/iaf_cond_alpha.h:
```diff
@@ -66,12 +66,14 @@ Simple conductance based leaky integrate-and-fire neuron model
 conductance-based synapses. Incoming spike events induce a postsynaptic change
 of conductance modelled by an alpha function. The alpha function
 is normalized such that an event of weight 1.0 results in a peak current of 1 nS
 at :math:`t = \tau_{syn}`.
```
This does the `t = \tau_{syn}` whereas above I corrected to double back-tics. Should probably also been `:math:` above, right?

terhorstd commented on models/iaf_psc_alpha.h:
```diff
@@ -68,7 +68,7 @@ Critical tests for the formulation of the neuron model are the
 comparisons of simulation results for different computation step
 sizes. sli/testsuite/nest contains a number of such tests.
```
```suggestion
sizes. The folder ``sli/testsuite/nest`` contains a number of such tests.
```

terhorstd commented on models/iaf_psc_delta_ps.h:
```diff
@@ -87,7 +87,7 @@ Critical tests for the formulation of the neuron model are the
 comparisons of simulation results for different computation step
 sizes. sli/testsuite/nest contains a number of such tests.
```
```suggestion
sizes. The folder ``sli/testsuite/nest`` contains a number of such tests.
```

terhorstd commented on models/iaf_psc_exp_multisynapse.h:
```diff
@@ -48,19 +48,20 @@ Leaky integrate-and-fire neuron model with multiple ports
 
 This can be reached by specifying separate receptor ports, each for
 a different time constant. The port number has to match the respective
 "receptor_type" in the connectors.
```
```suggestion
``receptor_type`` in the connectors.
```

terhorstd commented on models/iaf_psc_exp_ps_lossless.h:
```diff
@@ -268,7 +268,7 @@ class iaf_psc_exp_ps_lossless : public ArchivingNode
    * the final timestep
    * V > f(h, I) (or) linear approximation of the envelope, V < g(h, I_e).
-   * Note that in Algorithm 1 and 2 of [1], a typo interchanges g and f.
+   * Note that in Algorithm 1 and 2 of [1]_, a typo interchanges g and f.
```
No `:math:` in the equations above?

jessica-mitchell replied:
This is not part of the userdocs section, so the underscore is actually incorrect. Removing it.

terhorstd commented on models/rate_neuron_ipn.h:
```diff
@@ -74,15 +74,18 @@ or
- - mult_coupling_in (factor of multiplicative coupling for inhibitory input)
 
-The boolean parameter linear_summation determines whether the input function
+- input (nonlinearity that is applied to the input, either psi or phi)
+- mult_coupling_ex (factor of multiplicative coupling for excitatory input)
+- mult_coupling_in (factor of multiplicative coupling for inhibitory input)
```
These are the objects from the equation, do they need `:math:` or a number of back-ticks?
```suggestion
- `input` (nonlinearity that is applied to the input, either psi or phi)
- `mult_coupling_ex` (factor of multiplicative coupling for excitatory input)
- `mult_coupling_in` (factor of multiplicative coupling for inhibitory input)
```

terhorstd commented on models/rate_neuron_opn.h:
```diff
@@ -78,15 +78,18 @@ Here :math:`xi_j(t)` denotes a Gaussian white noise.
- - mult_coupling_in (factor of multiplicative coupling for inhibitory input)
 
-The boolean parameter linear_summation determines whether the input function
+- input (nonlinearity that is applied to the input, either psi or phi)
+- mult_coupling_ex (factor of multiplicative coupling for excitatory input)
+- mult_coupling_in (factor of multiplicative coupling for inhibitory input)
```
Same here

terhorstd commented on models/tanh_rate.h:
```diff
@@ -44,22 +44,23 @@ rate model with hyperbolic tangent non-linearity
 
 Nonlinear rate neurons can be created by typing
-nest.Create('tanh_rate_ipn') or nest.Create('tanh_rate_opn') for input
+``nest.Create('tanh_rate_ipn')`` or ``nest.Create('tanh_rate_opn')`` for input
```
Difficult to have pycodestyle check also the documentation parts, but for consistency:
```suggestion
``nest.Create("tanh_rate_ipn")`` or ``nest.Create("tanh_rate_opn")`` for input
```

terhorstd commented on models/tanh_rate.h:
```diff
@@ -44,22 +44,23 @@ rate model with hyperbolic tangent non-linearity
+``nest.Create('tanh_rate_ipn')`` or ``nest.Create('tanh_rate_opn')`` for input
 noise or output noise, respectively. Nonlinear rate transformers can
-be created by typing nest.Create('rate_transformer_tanh').
+be created by typing ``nest.Create('rate_transformer_tanh')``.
```
```suggestion
be created by typing ``nest.Create("rate_transformer_tanh")``.
```

terhorstd reviewed (changes requested):
This really fixes a lot of broken documentation! Much appreciated!!!
 :rocket: 

I have some small questions and comments just for clarification. Of course there are still tonnes of unquoted variables or math terms around, enough for such PRs in months to come. For this one we should however not take more changes into the set. Let's try to finish this one first, then eventually start a new one.

heplesser commented:
I just tried to build documentation from this branch. This failed because nest could not be imported with our fancy new way of managing the nest module in Python. Has anyone else seen this problem? The testsuite runs find and PyNEST works otherwise.

```
Running Sphinx v3.5.3

Configuration error:
There is a programmable error in your configuration file:

Traceback (most recent call last):
  File "/Users/plesser/miniforge3/envs/nest-simulator-dev/lib/python3.9/site-packages/sphinx/config.py", line 327, in eval_config_file
    execfile_(filename, namespace)
  File "/Users/plesser/miniforge3/envs/nest-simulator-dev/lib/python3.9/site-packages/sphinx/util/pycompat.py", line 88, in execfile_
    exec(code, _globals)
  File "/Users/plesser/NEST/bld_jessica/doc/userdoc/conf.py", line 87, in <module>
    nest.NestModule = type(nest)
  File "/Users/plesser/NEST/src/pynest/nest/__init__.py", line 423, in _setattr_error
    raise err from None
AttributeError: can't set attribute 'NestModule' on module 'nest'
```

jougs commented on models/tsodyks_synapse_hom.h:
```diff
@@ -57,25 +58,26 @@ to reproduce the results of [1]_ and to use this model of synaptic plasticity
 dynamics for their synaptic current or conductance. The effective synaptic
 weight, which will be transmitted to the postsynaptic neuron upon occurrence
-of a spike at time t is u(t)*x(t)*w, where u(t) and x(t) are defined in
+of a spike at time t is `u(t)*x(t)*w`, where u(t) and x(t) are defined in
 Eq (3) and (4), w is the synaptic weight specified upon connection.
```
```suggestion
of a spike at time t is :math:`u(t) \cdot x(t) \cdot w`, where `u(t)` and `x(t)` are defined in
Eq (3) and (4), `w` is the synaptic weight specified upon connection.
```

jougs commented on models/tsodyks_synapse_hom.h:
```diff
@@ -57,25 +58,26 @@ to reproduce the results of [1]_ and to use this model of synaptic plasticity
+of a spike at time t is `u(t)*x(t)*w`, where u(t) and x(t) are defined in
 Eq (3) and (4), w is the synaptic weight specified upon connection.
-The interpretation is as follows: The quantity u(t)*x(t) is the release
+The interpretation is as follows: The quantity `u(t)*x(t)` is the release
 probability times the amount of releasable synaptic vesicles at time t of the
```
```suggestion
The interpretation is as follows: The quantity :math:`u(t) \cdot x(t)` is the release
probability times the amount of releasable synaptic vesicles at time `t` of the
```

jougs commented on models/tsodyks_synapse.h:
```diff
@@ -59,29 +60,30 @@ to reproduce the results of [1]_ and to use this model of synaptic plasticity
-impulse u(t)*x(t)*w the full shape of y(t). The postsynaptic neuron, however,
+impulse `u(t)*x(t)*w` the full shape of y(t). The postsynaptic neuron, however,
 might choose to have a synaptic current that is not necessarily identical to
 the concentration of transmitter y(t) in the synaptic cleft. It may realize
 an arbitrary postsynaptic effect depending on y(t).
```
```suggestion
the concentration of transmitter `y(t)` in the synaptic cleft. It may realize
an arbitrary postsynaptic effect depending on `y(t)`.
```

jougs commented on models/tsodyks_synapse.h:
```diff
@@ -59,29 +60,30 @@ to reproduce the results of [1]_ and to use this model of synaptic plasticity
 into the synaptic cleft.
+
 The amount of transmitter than relaxes back to 0 with time constant tau_psc
 of the synapse's variable y. Since the dynamics of y(t) is linear, the
```
```suggestion
of the synapse's variable `y`. Since the dynamics of `y(t)` is linear, the
```

jougs reviewed (changes requested):
Many thanks! This is really a great improvement. Above my suggestions, I have a feeling that there might be more variable names in the tsodyks_* files that need marking up.

jougs commented:
@heplesser: This should be fixed by #2244.
@jessica-mitchell: can you please merge upstream/master?

jougs commented:
And just for the record: There is a certain potential for conflict with #2246.

jougs commented:
@jessica-mitchell: friendly ping!

jessica-mitchell commented:
@jougs thanks for the reminder -I am working on this - Sphinx failed to build after the changes you or Dennis made in the extractor_userdocs.py were committed, so I'm trying to fix it (I have a fix, but I've noticed another issue, which I'm looking into)

jessica-mitchell commented on models/gamma_sup_generator.h:
```diff
@@ -137,7 +140,7 @@ class gamma_sup_generator : public StimulationDevice
   struct Parameters_
   {
     double rate_;               //!< rate of component gamma process [Hz]
-    unsigned long gamma_shape_; //!< gamma shape parameter [1]_
+    unsigned long gamma_shape_; //!< gamma shape parameter [1]
```
yah I accidentally went too far with a couple of those into other comments,  should be fixed

jougs reviewed (approved):
The recent commits address all my concerns. Thanks again!
